cdap-authorization-ldap-role/src/main/java/io/cdap/cdap/security/authorization/ldap/role/searcher/LDAPSearcher.java

```
@@ -0,0 +1,147 @@
+import javax.naming.Context;
+import javax.naming.NamingEnumeration;
+import javax.naming.NamingException;
+import javax.naming.PartialResultException;
+import javax.naming.directory.Attribute;
+import javax.naming.directory.DirContext;
+import javax.naming.directory.SearchControls;
+import javax.naming.directory.SearchResult;
+
+import static javax.naming.directory.SearchControls.SUBTREE_SCOPE;
+
+
+/**
+ * Service for searching user's groups in LDAP
+ */
+public class LDAPSearcher {
+  private static final Logger LOG = LoggerFactory.getLogger(LDAPSearcher.class);
+
+  private static final String LDAP_SOCKET_FACTORY = "java.naming.ldap.factory.socket";
+  private static final String LDAPS_PROTOCOl = "ldaps";
+  private static final String BASE_DN_SPLITTER = ";";
+  private static final int MAX_RETRIES = 5;
+
+  private final LDAPSearchConfig config;
+  private final String[] baseDNList;
+
+  public LDAPSearcher(LDAPSearchConfig config) {
+    this.config = config;
+    baseDNList = config.getSearchBaseDn().split(BASE_DN_SPLITTER);
+  }
+
+  public void testConnection() {
+    try {
+      DirContext context = getConnection();
+      context.close();
+    } catch (NamingException e) {
+      LOG.error("Failed to establish connection to '{}'", config.getUrl(), e);
+      throw new RuntimeException(e);
+    }
+  }
+
+  public Set<String> searchGroups(String username) {
+    Set<String> groups = Collections.emptySet();
+
+    try {
+      DirContext context = getConnection();
+
+      SearchControls controls = new SearchControls();
+      if (config.isRecursiveSearch()) {
+        controls.setSearchScope(SUBTREE_SCOPE);
+      }
+
+      groups = Arrays.stream(baseDNList)
+        .map(baseDN -> searchGroups(baseDN, username, context, controls))
+        .flatMap(Collection::stream)
+        .collect(Collectors.toSet());
+
+      context.close();
+    } catch (NamingException e) {
+      LOG.error("Failed to find groups for '{}'", username, e);
```

definitely should not just ignore an exception. Also what about other exceptions?

---

done